### PR TITLE
fix(HMS-1403): timeout for job queue

### PR DESCRIPTION
**Pull request by lzap**
This implements a relatively long (default) timeout for job queue. The purpose of this is when there is a bug that makes jobs to get stuck, resources (CPU/memory) can be kept on a good value until we figure out what is going on. I propose 30 minutes as a safe timeout value.

The implementation uses context deadline and adds checks to every job queue "step" (function) which is using the reservation status update so jobs are cancelled early. Of course, both AWS and Azure SDKs do support context cancellation, so in practice, job should be cancelled immediately once it reaches the deadline. The `nilUnlessDeadline` function serves a safety measure in case we copy-paste existing code and we use a library that does not support context deadlines.

Initially, I used `panic` to handle timeouts but this was very hard to test. This new version use explicit checking via the newly introduced function. The PR is separated into three commits:

* the timeout implementation
* new integration test and refactoring needed in order to be able to introduce Redis
* enforcement of config and make documentation via CI job @ezr-ondrej you really wanted this

### CODE REVIEW

**Comment by lzap** [comment]
Do not merge, using `panic` is causing serious issues in tests.

**Comment by lzap** [comment]
Finally this is ready.

**Comment by adiabramovitch** [comment]
@lzap SOmething weird why are the tests still running?

**Comment by lzap** [comment]
That is sort of expected because I renamed "Go tests" to more informative "Unit tests". Also renamed "DAO tests" to "Integration tests". Not sure why "Go tests" triggered while "DAO tests" already disappeared. To to double check, run them locally please.

**Comment by ezr-ondrej** [comment]
> Also renamed "DAO tests" to "Integration tests". Not sure why "Go tests" triggered while "DAO tests" already disappeared.

Both are there (under scroll bar), because both are required checks, will change it once this is merged :+1:

File: mk/z_common.mk
Comment (ezr-ondrej) [inline mk/z_common.mk:6]: :orange_heart:

File: pkg/worker/redis.go
Comment (lzap) [inline pkg/worker/redis.go:216]: These three lines are actually the change + configuration option.

All the rest are just tests :-)

File: pkg/worker/redis.go
Comment (lzap) [inline pkg/worker/redis.go:164]: Ah this is not even needed, the initial version contained some changes in this function but I dropped it (it was really hard to test for `panic`). I can remove this hunk.

File: internal/testing/integration/db_env.go
Comment (ezr-ondrej) [inline internal/testing/integration/db_env.go:21]: Probably a leftover?

File: pkg/worker/redis.go
Comment (ezr-ondrej) [reply pkg/worker/redis.go:216]: Shall we add this to memory implementation as well?
You know, just for consistency...

File: internal/queue/tests/main_test.go
Comment (ezr-ondrej) [inline internal/queue/tests/main_test.go:29]: I'm really nitpicking here, pls feel free to ignore me, but this would look nice in a function (`configureJobqueue` for example).
But I'm all for merging as is and addressing this in a follow-up, if I get around to this, such a global setting would look nice in a config file, that we would create for the GH actions (definitelly out of scope this PR).

**Review by ezr-ondrej — COMMENTED** [review]
Very nice job indeed, I love the tests :orange_heart: and basically all about this PR :+1:
I've got just one comment I'd like to see addressed, I think we have a leftover there) all the other comments are not blocking :+1:

**Comment by ezr-ondrej** [comment]
I'll force merge once ready btw and rename the required checks right after :+1:

File: pkg/worker/redis.go
Comment (lzap) [reply pkg/worker/redis.go:164]: Yes this was a leftover, I will amend this change which just improves the error handling - when the panic was caused by error we use the native stack tracing and error reporting of zerolog. Otherwise we just do manual stacktrace:

```
		if err, ok := rec.(error); ok {
			logger.Err(err).Stack().Msgf("Job queue panic: %s", err.Error())
		} else {
			logger.Msgf("Error during job handling: %v, stacktrace: %s", rec, debug.Stack())
		}
```

File: pkg/worker/redis.go
Comment (lzap) [reply pkg/worker/redis.go:216]: Memory is only used for development, I don’t see any benefit but why not.

File: internal/queue/tests/main_test.go
Comment (lzap) [reply internal/queue/tests/main_test.go:29]: Don’t exactly understand what you mean.

**Comment by lzap** [comment]
Pushed changes.

**Comment by ezr-ondrej** [comment]
The new validation is failing as GCP PR had broken it :(

**Comment by lzap** [comment]
What was failing? Docs? I could have fixed that, but thanks.

**Comment by lzap** [comment]
NP did not notice it.

**Review by ezr-ondrej — APPROVED** [review]
Thanks ! :orange_heart:

**Pull request merged**
